src/vos/vos_aggregate.c

```diff
@@ -195,20 +442,997 @@ vos_agg_sv(daos_handle_t ih, vos_iter_entry_t *entry,
+	D_ASSERT(lgc_seg->ls_idx_start <= lgc_seg->ls_idx_end);
+	D_ASSERT(lgc_seg->ls_idx_end < mw->mw_lgc_cnt);
+
+	bio_ctxt = obj->obj_cont->vc_pool->vp_io_ctxt;
+	D_ASSERT(bio_ctxt != NULL);
+
+	iov.iov_buf = io->ic_buf;
+	iov.iov_buf_len = io->ic_buf_len;
+
+	i = lgc_seg->ls_idx_start;
+	while (i <= lgc_seg->ls_idx_end) {
+		if (lgc_seg->ls_phy_ent != NULL) {
+			phy_ent = lgc_seg->ls_phy_ent;
+			ext = ent_in->ei_rect.rc_ex;
+		} else {
+			struct agg_lgc_ent *lgc_ent = &mw->mw_lgc_ents[i];
+
+			phy_ent = lgc_ent->le_phy_ent;
+			ext = lgc_ent->le_ext;
+		}
+		i++;
+
+		D_ASSERT(ext1_covers_ext2(&ent_in->ei_rect.rc_ex, &ext));
+		D_ASSERT(ext1_covers_ext2(&phy_ent->pe_rect.rc_ex, &ext));
+
+		phy_lo = phy_ent->pe_rect.rc_ex.ex_lo;
+		if (phy_ent->pe_off != 0)
+			phy_lo += phy_ent->pe_off;
+
+		D_ASSERT(phy_lo <= phy_ent->pe_rect.rc_ex.ex_hi);
+		D_ASSERT(ext.ex_lo >= phy_lo);
+
+		copy_size = evt_extent_width(&ext) * ent_in->ei_inob;
+
+		addr_src = phy_ent->pe_addr;
+		addr_src.ba_off += (ext.ex_lo - phy_lo) * ent_in->ei_inob;
+
+		D_ASSERT(!bio_addr_is_hole(&addr_src));
+		D_ASSERT(iov.iov_buf_len >= copy_size);
+
+		mark_yield(&addr_src, acts);
+		/*
+		 * Set 'iov_len' beforehand, cause it will be used as copy
+		 * size in bio_readv().
+		 */
+		iov.iov_len = copy_size;
+		rc = bio_readv(bio_ctxt, addr_src, &iov);
+		if (rc) {
+			D_ERROR("Read "DF_RECT" "DF_EXT" error: %d\n",
+				DP_RECT(&phy_ent->pe_rect), DP_EXT(&ext), rc);
+			return rc;
+		}
+		D_ASSERT(iov.iov_len == copy_size);
+
+		D_ASSERT(iov.iov_buf_len >= copy_size);
+		iov.iov_buf_len -= copy_size;
+		iov.iov_buf += copy_size;
+	}
+	D_ASSERT(seg_size == (io->ic_buf_len - iov.iov_buf_len));
+
+	addr_dst = ent_in->ei_addr;
+	D_ASSERT(!bio_addr_is_hole(&addr_dst));
+	mark_yield(&addr_dst, acts);
+
+	iov.iov_buf = io->ic_buf;
+	iov.iov_buf_len = io->ic_buf_len;
+	iov.iov_len = seg_size;
+	rc = bio_writev(bio_ctxt, addr_dst, &iov);
+	if (rc)
+		D_ERROR("Write "DF_RECT" "DF_EXT" error: %d\n",
+			DP_RECT(&ent_in->ei_rect), DP_EXT(&ext), rc);
+
+	return rc;
+}
+
+static int
+fill_segments(daos_handle_t ih, struct agg_merge_window *mw,
+	      unsigned int *acts)
+{
+	struct agg_io_context	*io = &mw->mw_io_ctxt;
+	struct agg_lgc_seg	*lgc_seg;
+	unsigned int		 i, scm_max;
+	int			 rc = 0;
+
+	scm_max = MAX(io->ic_seg_cnt, 200);
+	if (io->ic_scm_max < scm_max) {
+		struct pobj_action *scm_exts;
+
+		D_REALLOC(scm_exts, io->ic_scm_exts,
+			  scm_max * sizeof(*scm_exts));
+		if (scm_exts == NULL)
+			return -DER_NOMEM;
+
+		io->ic_scm_exts = scm_exts;
+		io->ic_scm_max = scm_max;
+	}
+	memset(io->ic_scm_exts, 0, io->ic_scm_max * sizeof(*io->ic_scm_exts));
+	D_ASSERT(io->ic_scm_cnt == 0);
+
+	for (i = 0; i < io->ic_seg_cnt; i++) {
+		lgc_seg = &io->ic_segs[i];
+
+		D_DEBUG(DB_EPC, "Fill segment: %u-%u "DF_RECT"\n",
+			lgc_seg->ls_idx_start, lgc_seg->ls_idx_end,
+			DP_RECT(&lgc_seg->ls_ent_in.ei_rect));
+
+		rc = fill_one_segment(ih, mw, lgc_seg, acts);
```

Comment:
in this loop, we will read+write without parallelism right? I'm a little concerned because we only have one ULT for aggregation and this ULT has to wait for SSD read+write latency before it can proceed?

We could improve those functions and have asynchronous version so we can post all reads, and start to post writes after that. Another option could be create a ULT for aggregation of each object, but it will have other overhead.

Reply:
Yes, I considered to improve it by issuing multiple blob read/write, that requires revise the bio_readv/writev() interface (or introduce a new set of bio interface), let's do it in follow-on PR.